SpectroscopicAbsorption QoI

tradowsk opened the pull request:
New QoI class for calculating laser absorption based on the Beer-Lambert Law.

The HITRAN class reads data_files from the HITRAN Spectroscopic Database and makes the data available for use in the Absorbance class.

The Absorbance class evaluates the Beer-Lambert Law at a given point in space. This is intended to be supplied to the IntegratedFunction class for qoi evaluation. Note that this calculates the absorbance, alpha. To calculate laser absorption, I/I0 = exp(-alpha), so the SpectroscopicAbsorption class overrides `parallel_op()` and `thread_join()` to perform this final step.

CPPUnit tests are also provided for the HITRAN and SpectroscopicAbsorption classes. Calculated answer for SpectroscopicAbsorption is based upon a MATLAB code.

I post this now mainly for code comments. I cannot test with latest libMesh due to #460, and I am currently seeking a third-party validation of the QoI calculated answer.

pbauman commented:
I'll take a closer look once you pull the latest GRINS and start fixing all the compiling errors you should be getting.

tradowsk commented:
Rebased, and all tests pass. No compiler issues for me other than #460.

pbauman commented on src/qoi/include/grins/hitran.h:
```diff
@@ -0,0 +1,154 @@
+      @param data_file A file containing values for the require quantities, formatted as described above
+      @param parFunction_file A file containing temperature and partition function values, formatted as described above
+    */
+    HITRAN(const std::string& data_file, const std::string& parFunction_file);
```
I've started doing the libMesh spacing for `*` and `&`, e.g. `const std::string &`. Let's do that here too, please.

Also, camel case can die in a fire. Please use all lower case for variable and functionnames, with a `_` separating keywords.

pbauman commented on src/qoi/include/grins/hitran.h:
```diff
@@ -0,0 +1,154 @@
+    int _T_index(libMesh::Real T);
+
+    //! Search through the partition function data to get value for given temperature
+    libMesh::Real search_parFunction(libMesh::Real T);
```
No camel case.

pbauman commented:
Could we also include your `SpectroscopicAbsorption` test as an example (perhaps to be updated later with a more interesting flow) where you show how to use the HITRAN-Python foo to generate the HITRAN data and use it through GRINS?

tradowsk commented:
My python code requires the use of HAPI to actually access the HITRAN data. Would you want this library included in the example? It requires python3

pbauman commented:
Da fuq? Not a fan of the review feature. Thought all those comments were going live as I hit enter.

pbauman commented:
> My python code requires the use of HAPI to actually access the HITRAN data. Would you want this library included in the example? It requires python3

I think it's enough to include a README on the prereqs and a working Python script that would work with the prereqs satisfied. I'm indifferent to the flavor of Python - this is more about documenting for the future about how to get data for this new feature.

tradowsk commented:
> Da fuq? Not a fan of the review feature. Thought all those comments were going live as I hit enter.

Apparently not...

I can update my fixes and rebase again

pbauman commented:
Just squash in the fixes and let me know when you've finished them. Then will have a look again.

tradowsk commented on src/utilities/src/string_utils.C:
```diff
@@ -49,5 +49,27 @@ namespace GRINS
+                            const std::string& delimiter,
+                            std::vector<libMesh::Real>& results )
+    {
+      // Skip delimiters at beginning.
```
But then I need to create a temporary vector of type `string` to pass to `split_string()`, and then convert and copy all the values from that to `results`. Seems inefficient to me. Or did you have a different idea?

tradowsk commented on src/qoi/src/absorbance.C:
```diff
@@ -0,0 +1,236 @@
+    _species_idx = _chemistry->species_index(species); 
+    _nu0 = _hitran->nu0();
+    
+    if (thermo_pressure == -1.0) {
```
`@param termo_pressure The thermodynamic pressure (in [Pa]), or -1.0 if non-constant`

It was just a dummy, non-physical value to use as a flag for whether or not the thermopressure would be calculated or provided explicitly in the input file.

tradowsk replied:
I guess I can overload the constructor, and have the thermo_pressure missing from the parameters if it is non-constant

tradowsk commented on src/qoi/src/spectroscopic_absorption.C:
```diff
@@ -0,0 +1,74 @@
+namespace GRINS
+{
+  SpectroscopicAbsorption::SpectroscopicAbsorption( const std::string& qoi_name, SharedPtr<libMesh::FEMFunctionBase<libMesh::Real> > absorb, SharedPtr<RayfireMesh> rayfire)
+    : IntegratedFunction<libMesh::FEMFunctionBase<libMesh::Real> >(2,absorb,rayfire,qoi_name)
```
That 2 is the quadrature order used in `IntegratedFunction`

tradowsk commented on src/qoi/src/qoi_factory.C:
```diff
@@ -149,11 +135,60 @@ namespace GRINS
+        else
+          libmesh_error_msg("ERROR: Could not find material parameter");
+      
+        std::string hitran_data, hitran_parFunction;
```
He says in the camel cased `QoIFactory` :-p

tradowsk commented:
Squashed in the syntax fixes, for some reason GitHub displays the commits in a different order than on my `git log`.

When I rebase to fix `string_utils` it should fix it

tradowsk commented:
Squashed in the string_utils fix. Still can't figure out why GitHub commit ordering is different than what I can see on my `git log`. Any ideas? Or do I just keep rebasing until I get lucky and get the right order...

tradowsk commented:
While adding "moar mathz," I found that my choice of `Absorbance` for the class name is misleading, since what is calculated in that class is actually the _absorption coefficient_ k_{\nu}, while the absorbance is k_{\nu)*L. I updated the class name appropriately.

tradowsk commented:
Okay, I refactored and rebased my code to include functionality to handle multiple spectroscopic lines and isotopologues in calculating the absorption coefficient

pbauman commented on src/qoi/src/absorption_coeff.C:
```diff
@@ -0,0 +1,290 @@
+      _rad_coeff(1.43887752) // [cm K]
+  {
+    // sanity checks
+    libmesh_assert_greater(nu_max,nu_min);
```
I would actually make these runtime checks/errors, not just debug/devel mode checks.

pbauman commented on src/qoi/src/hitran.C:
```diff
@@ -0,0 +1,247 @@
+                 libMesh::Real T_min, libMesh::Real T_max, libMesh::Real T_step)
+  {
+    // sanity checks on temperature range specification
+    libmesh_assert_greater(T_min,0.0);
```
Non-assert check/errors too please.

pbauman commented on src/qoi/src/hitran.C:
```diff
@@ -0,0 +1,247 @@
+
+    if (i >= 0)
+      {
+        if (_T[i] == T)
```
This is dangerous. Should use a tolerance based check instead.

pbauman commented on src/qoi/src/spectroscopic_absorption.C:
```diff
@@ -0,0 +1,74 @@
+
+    // absorbance is calculated in [cm^-1], but path length is given in [m]
+    // 100.0 factor converts pathlength to [cm]
+    QoIBase::_qoi_value = std::exp( -QoIBase::_qoi_value * 100.0 );
```
You should be setting the value of qoi here.

tradowsk commented on src/qoi/src/hitran.C:
```diff
@@ -0,0 +1,247 @@
+    return retval;
+  }
+
+  int HITRAN::_T_index(libMesh::Real T)
```
Isn't `std::find()` just a linear search though?

tradowsk commented on src/qoi/src/spectroscopic_absorption.C:
```diff
@@ -0,0 +1,74 @@
+
+    // absorbance is calculated in [cm^-1], but path length is given in [m]
+    // 100.0 factor converts pathlength to [cm]
+    QoIBase::_qoi_value = std::exp( -QoIBase::_qoi_value * 100.0 );
```
I think I figured out how this works, but I wanted to double check. For my tests, it seems that `thread_join` is called, and then `parallel_op`. Is this always the case? Because then I can just leave `thread_join` alone to do a sum, and use `parallel_op` to do the exponentiation.

tradowsk commented:
Passing `make check` on `dbg`,`devel`,&`opt`, and I think that I covered everything in your code comments.

pbauman commented:
Looks like this should be good to merge. Thanks!
